libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3816,6 +3816,20 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
 				)
 			);
 
+		if (
+			functionDefinition.visibility() == Visibility::Private &&
+			functionDefinition.scope() != m_currentContract
+		)
+			m_errorReporter.typeError(
+				6772_error,
+				path->location(),
+				fmt::format(
+					"Function \"{}\" is private and therefore cannot be attached"
+					" to a type outside of the library where it is defined.",
+					joinHumanReadable(path->path(), ".")
+				)
```

Comment:
I'd add an assert inside the if that the function is a library function. If we ever allow non-contract functions to be private, the message will need to be reworded.